modules/atomic-widgets/base/atomic-element-base.php

```diff
@@ -0,0 +1,127 @@
+<?php
+
+namespace Elementor\Modules\AtomicWidgets\Base;
+
+use Elementor\Element_Base;
+use Elementor\Modules\AtomicWidgets\Controls\Section;
+use Elementor\Modules\AtomicWidgets\PropTypes\Concerns\Has_Atomic_Base;
+use Elementor\Modules\AtomicWidgets\Styles\Style_Schema;
+use Elementor\Modules\AtomicWidgets\Validators\Props_Validator;
+use Elementor\Modules\AtomicWidgets\PropTypes\Contracts\Prop_Type;
+use Elementor\Modules\AtomicWidgets\Validators\Styles_Validator;
+use Elementor\Utils;
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+abstract class Atomic_Element_Base extends Element_Base {
+
+	use Has_Atomic_Base;
+
+	protected $version = '0.0';
+	protected $styles = [];
+
+	public function __construct( $data = [], $args = null ) {
+		parent::__construct( $data, $args );
+
+		$this->version = $data['version'] ?? '0.0';
+		$this->styles = $data['styles'] ?? [];
+	}
+
+	private function get_valid_controls( array $schema, array $controls ): array {
+		$valid_controls = [];
+
+		foreach ( $controls as $control ) {
+			if ( $control instanceof Section ) {
+				$cloned_section = clone $control;
+
+				$cloned_section->set_items(
+					$this->get_valid_controls( $schema, $control->get_items() )
+				);
+
+				$valid_controls[] = $cloned_section;
+				continue;
+			}
+
+			if ( ! ( $control instanceof Atomic_Control_Base ) ) {
+				Utils::safe_throw( 'Control must be an instance of `Atomic_Control_Base`.' );
+				continue;
+			}
+
+			$prop_name = $control->get_bind();
+
+			if ( ! $prop_name ) {
+				Utils::safe_throw( 'Control is missing a bound prop from the schema.' );
+				continue;
+			}
+
+			if ( ! array_key_exists( $prop_name, $schema ) ) {
+				Utils::safe_throw( "Prop `{$prop_name}` is not defined in the schema of `{$this->get_name()}`." );
+				continue;
+			}
+
+			$valid_controls[] = $control;
+		}
+
+		return $valid_controls;
+	}
+
+	abstract protected function define_atomic_controls(): array;
+
+
+	final public function get_initial_config() {
+		$config = parent::get_initial_config();
+
+		$config['atomic_controls'] = $this->get_atomic_controls();
+		$config['atomic_props_schema'] = static::get_props_schema();
+		$config['version'] = $this->version;
+		$config['show_in_panel'] = true;
+		$config['categories'] = [ 'layout' ];
+		$config['hide_on_search'] = false;
+		$config['controls'] = [];
+
+		return $config;
+	}
+
+	private static function validate_schema( array $schema ) {
```

Comment:
```
validate_schema
sanitize_atomic_styles
sanitize_atomic_settings
```

Can any of these also be moved to the trait?

Reply:
@baghdasarovelementor yes/no?

Reply:
@louiswol94 Moved